File: module4_dr_grading python/src/prepareDataset.m
function [trainDS, valDS, testDS] = prepareDataset(dataFolder, targetSize, splitRatios, mode)
% PREPAREDATASET Sets up imageDatastores, performs patient-level splitting 
% to prevent data leakage, applies image preprocessing, and combines them
% with CORAL-encoded ordinal targets and optional clinical features.
%
% Inputs:
%   dataFolder  - Path to the dataset folder
%   targetSize  - Image target dimensions (default: [224, 224])
%   splitRatios - 1x3 vector for train/val/test splits (default: [0.7, 0.15, 0.15])
%   mode         - String: 'cnn-only', 'lesion-only', or 'fusion' (default: 'fusion')
% Outputs:
%   trainDS, valDS, testDS - Combined datastores returning:
%       - 'cnn-only':    {preprocessedImage, cumulativeTarget}
%       - 'lesion-only': {lesionVector, cumulativeTarget}
%       - 'fusion':      {preprocessedImage, lesionVector, cumulativeTarget}

    if nargin < 2 || isempty(targetSize)
        targetSize = [224, 224];
    end
    if nargin < 3 || isempty(splitRatios)
        splitRatios = [0.7, 0.15, 0.15];
    end
    if nargin < 4 || isempty(mode)
        mode = 'fusion';
    end
    
    % Normalize split ratios to sum to 1.0
    splitRatios = splitRatios / sum(splitRatios);

    % Create the base imageDatastore
    imds = imageDatastore(dataFolder, ...
        'IncludeSubfolders', true, ...
        'LabelSource', 'foldernames');
    
    files = imds.Files;
    nFiles = numel(files);
    patientIDs = cell(nFiles, 1);
    
    % Extract patient IDs from filenames to prevent leakage
    for k = 1:nFiles
        patientIDs{k} = getPatientID(files{k});
    end
    
    [uniquePatients, ~, idxMap] = unique(patientIDs);
    nPatients = numel(uniquePatients);
    
    % Deterministic random shuffle of patients
    rng(42);
    shuffledPatients = randperm(nPatients);
    
    % Calculate patient-level splits
    nTrain = round(splitRatios(1) * nPatients);
    nVal = round(splitRatios(2) * nPatients);
    
    trainPatientsIdx = shuffledPatients(1:nTrain);
    valPatientsIdx = shuffledPatients(nTrain+1 : nTrain+nVal);
    testPatientsIdx = shuffledPatients(nTrain+nVal+1 : end);
    
    % Map patient splits back to image indices
    trainIdx = ismember(idxMap, trainPatientsIdx);
    valIdx = ismember(idxMap, valPatientsIdx);
    testIdx = ismember(idxMap, testPatientsIdx);
    
    % Construct individual datastores
    trainImds = subset(imds, trainIdx);
    valImds = subset(imds, valIdx);
    testImds = subset(imds, testIdx);
    
    % Create cumulative CORAL targets for each subset (size N x 4)
    trainTargets = makeCoralTargets(trainImds.Labels);
    valTargets = makeCoralTargets(valImds.Labels);
    testTargets = makeCoralTargets(testImds.Labels);
    
    % Create target arrayDatastores and transpose to [4, 1] column vectors
    trainTargetDS = transform(arrayDatastore(trainTargets), @(t) t');
    valTargetDS = transform(arrayDatastore(valTargets), @(t) t');
    testTargetDS = transform(arrayDatastore(testTargets), @(t) t');
    
    % Assemble datastores based on selected mode
    if strcmp(mode, 'cnn-only')
        % Image-only pipeline
        trainImdsTrans = transform(trainImds, @(img) preprocessImage(img));
        valImdsTrans = transform(valImds, @(img) preprocessImage(img));
        testImdsTrans = transform(testImds, @(img) preprocessImage(img));
        
        trainDS = combine(trainImdsTrans, trainTargetDS);
        valDS = combine(valImdsTrans, valTargetDS);
        testDS = combine(testImdsTrans, testTargetDS);
        
    elseif strcmp(mode, 'lesion-only')
        % Lesion-only pipeline
        trainLesions = makeMockLesions(trainImds.Labels);
        valLesions = makeMockLesions(valImds.Labels);
        testLesions = makeMockLesions(testImds.Labels);
        
        % Create lesion arrayDatastores and transpose to [8, 1] column vectors
        trainLesionDS = transform(arrayDatastore(trainLesions), @(l) l');
        valLesionDS = transform(arrayDatastore(valLesions), @(l) l');
        testLesionDS = transform(arrayDatastore(testLesions), @(l) l');
        
        trainDS = combine(trainLesionDS, trainTargetDS);
        valDS = combine(valLesionDS, valTargetDS);
        testDS = combine(testLesionDS, testTargetDS);
        
    else % 'fusion'
        % Combined dual-stream pipeline
        trainImdsTrans = transform(trainImds, @(img) preprocessImage(img));
        valImdsTrans = transform(valImds, @(img) preprocessImage(img));
        testImdsTrans = transform(testImds, @(img) preprocessImage(img));
        
        trainLesions = makeMockLesions(trainImds.Labels);
        valLesions = makeMockLesions(valImds.Labels);
        testLesions = makeMockLesions(testImds.Labels);
        
        % Create lesion arrayDatastores and transpose to [8, 1] column vectors
        trainLesionDS = transform(arrayDatastore(trainLesions), @(l) l');
        valLesionDS = transform(arrayDatastore(valLesions), @(l) l');
        testLesionDS = transform(arrayDatastore(testLesions), @(l) l');
        
        trainDS = combine(trainImdsTrans, trainLesionDS, trainTargetDS);
        valDS = combine(valImdsTrans, valLesionDS, valTargetDS);
        testDS = combine(testImdsTrans, testLesionDS, testTargetDS);
    end
end

function patientID = getPatientID(filepath)
    % Extracts patient ID from a file path to prevent leakage of bilateral scans
    [~, filename, ~] = fileparts(filepath);
    
    % Pattern 1: Messidor-2 style (e.g., 20051020_43808_0100_PP.png -> 43808)
    tokens = regexp(filename, '^\d+_(\d+)_\d+_PP$', 'tokens');
    if ~isempty(tokens)
        patientID = tokens{1}{1};
        return;
    end
    
    % Pattern 2: Simple prefix split by underscore (e.g., 123_left.png -> 123)
    tokens = regexp(filename, '^([^_]+)_', 'tokens');
    if ~isempty(tokens)
        patientID = tokens{1}{1};
        return;
    end
    
    % Fallback: Treat each image name as a unique patient
    patientID = filename;
end

function targets = makeCoralTargets(labels)
    % Encodes categorical labels into CORAL binary cumulative vectors.
    nImages = numel(labels);
    nClasses = 5;
    targets = zeros(nImages, nClasses - 1, 'single');
    for k = 1:nImages
        grade = double(string(labels(k)));
        if grade > 0
            targets(k, 1:grade) = 1.0;
        end
    end
end

function lesions = makeMockLesions(labels)
    % Generates synthetic lesion feature vectors mapped to ground-truth grades.
    % Injects 10% failure rates to train robust fallback nodes.
    nImages = numel(labels);
    lesions = zeros(nImages, 8, 'single');
    
    % Reset seed locally to ensure reproducibility
    rng(100);
    
    for k = 1:nImages
        grade = double(string(labels(k)));
        v = zeros(1, 8, 'single');
        v(7) = 1.0; % I_avail = 1.0
        v(8) = 0.0; % I_missing = 0.0
        
        switch grade
            case 0
                v(1) = log1p(randi([0, 1])); % Microaneurysms
                v(2) = rand() * 0.01;        % Hemorrhage area
                v(3) = rand() * 0.01;        % Exudate area
                v(4) = 0.5 + rand() * 0.2;   % Vessel density
                v(5) = 0.0;                  % NV score
                v(6) = 0.8 + rand() * 0.2;   % Optic disc distance
            case 1
                v(1) = log1p(randi([1, 5]));
                v(2) = rand() * 0.02;
                v(3) = rand() * 0.01;
                v(4) = 0.5 + rand() * 0.2;
                v(5) = 0.0;
                v(6) = rand() * 0.8;
            case 2
                v(1) = log1p(randi([5, 15]));
                v(2) = rand() * 0.05;
                v(3) = rand() * 0.08;
                v(4) = 0.4 + rand() * 0.2;
                v(5) = 0.0;
                v(6) = rand() * 0.7;
            case 3
                v(1) = log1p(randi([15, 40]));
                v(2) = 0.05 + rand() * 0.15;
                v(3) = 0.08 + rand() * 0.15;
                v(4) = 0.3 + rand() * 0.2;
                v(5) = rand() * 0.05;
                v(6) = rand() * 0.5;
            case 4
                v(1) = log1p(randi([30, 80]));
                v(2) = 0.10 + rand() * 0.25;
                v(3) = 0.10 + rand() * 0.25;
                v(4) = 0.2 + rand() * 0.2;
                v(5) = 0.1 + rand() * 0.4;
                v(6) = rand() * 0.4;
        end
        
        % 10% probability of M3 module failure simulation
        if rand() < 0.10
            v = zeros(1, 8, 'single');
            v(8) = 1.0; % I_missing = 1.0, I_avail = 0.0
        end
        
        lesions(k, :) = v;
    end
end

File: module4_dr_grading python/src/prepare_dataset.py
import os
import glob
import re
import numpy as np
import cv2
import torch
from torch.utils.data import Dataset, DataLoader, Subset
import sys

# Import preprocess_image
sys.path.append(os.path.join(os.path.dirname(__file__), 'utils'))
from preprocess_image import preprocess_image

def get_patient_id(filepath):
    filename = os.path.basename(filepath)
    name, _ = os.path.splitext(filename)
    
    # Pattern 1: Messidor-2 style
    m1 = re.match(r'^\d+_(\d+)_\d+_PP$', name)
    if m1:
        return m1.group(1)
        
    # Pattern 2: Simple prefix split by underscore
    m2 = re.match(r'^([^_]+)_', name)
    if m2:
        return m2.group(1)
        
    return name

def make_coral_targets(labels):
    n_images = len(labels)
    n_classes = 5
    targets = np.zeros((n_images, n_classes - 1), dtype=np.float32)
    for k in range(n_images):
        grade = int(labels[k])
        if grade > 0:
            targets[k, 0:grade] = 1.0
    return targets

def make_mock_lesions(labels):
    np.random.seed(100) # Reset seed locally to ensure reproducibility
    n_images = len(labels)
    lesions = np.zeros((n_images, 8), dtype=np.float32)
    
    for k in range(n_images):
        grade = int(labels[k])
        v = np.zeros(8, dtype=np.float32)
        v[6] = 1.0 # I_avail
        v[7] = 0.0 # I_missing
        
        if grade == 0:
            v[0] = np.log1p(np.random.randint(0, 2))
            v[1] = np.random.rand() * 0.01
            v[2] = np.random.rand() * 0.01
            v[3] = 0.5 + np.random.rand() * 0.2
            v[4] = 0.0
            v[5] = 0.8 + np.random.rand() * 0.2
        elif grade == 1:
            v[0] = np.log1p(np.random.randint(1, 6))
            v[1] = np.random.rand() * 0.02
            v[2] = np.random.rand() * 0.01
            v[3] = 0.5 + np.random.rand() * 0.2
            v[4] = 0.0
            v[5] = np.random.rand() * 0.8
        elif grade == 2:
            v[0] = np.log1p(np.random.randint(5, 16))
            v[1] = np.random.rand() * 0.05
            v[2] = np.random.rand() * 0.08
            v[3] = 0.4 + np.random.rand() * 0.2
            v[4] = 0.0
            v[5] = np.random.rand() * 0.7
        elif grade == 3:
            v[0] = np.log1p(np.random.randint(15, 41))
            v[1] = 0.05 + np.random.rand() * 0.15
            v[2] = 0.08 + np.random.rand() * 0.15
            v[3] = 0.3 + np.random.rand() * 0.2
            v[4] = np.random.rand() * 0.05
            v[5] = np.random.rand() * 0.5
        elif grade == 4:
            v[0] = np.log1p(np.random.randint(30, 81))
            v[1] = 0.10 + np.random.rand() * 0.25
            v[2] = 0.10 + np.random.rand() * 0.25
            v[3] = 0.2 + np.random.rand() * 0.2
            v[4] = 0.1 + np.random.rand() * 0.4
            v[5] = np.random.rand() * 0.4
            
        # 10% probability of M3 module failure simulation
        if np.random.rand() < 0.10:
            v = np.zeros(8, dtype=np.float32)
            v[7] = 1.0
            
        lesions[k, :] = v
        
    return lesions

class DRDataset(Dataset):
    def __init__(self, filepaths, labels, mode='fusion'):
        self.filepaths = filepaths
        self.labels = labels
        self.mode = mode
        
        self.targets = make_coral_targets(self.labels)
        if self.mode in ['lesion-only', 'fusion']:
            self.lesions = make_mock_lesions(self.labels)
            
    def __len__(self):
        return len(self.filepaths)
        
    def __getitem__(self, idx):
        target = torch.tensor(self.targets[idx], dtype=torch.float32)
        
        if self.mode == 'lesion-only':
            lesion = torch.tensor(self.lesions[idx], dtype=torch.float32)
            return lesion, target
            
        # Load image (OpenCV returns BGR, convert to RGB)
        img_bgr = cv2.imread(self.filepaths[idx])
        img_rgb = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)
        img_preprocessed = preprocess_image(img_rgb)
        # Convert HWC to CHW for PyTorch
        img_tensor = torch.tensor(img_preprocessed.transpose(2, 0, 1), dtype=torch.float32)
        
        if self.mode == 'cnn-only':
            return img_tensor, target
            
        # fusion
        lesion = torch.tensor(self.lesions[idx], dtype=torch.float32)
        return img_tensor, lesion, target

def prepare_dataset(data_folder, split_ratios=(0.7, 0.15, 0.15), mode='fusion'):
    # Find all images
    filepaths = []
    labels = []
    
    for c in range(5):
        class_folder = os.path.join(data_folder, str(c))
        if os.path.isdir(class_folder):
            files = glob.glob(os.path.join(class_folder, '*.png')) + glob.glob(os.path.join(class_folder, '*.jpg'))
            filepaths.extend(files)
            labels.extend([c] * len(files))
            
    n_files = len(filepaths)
    patient_ids = [get_patient_id(f) for f in filepaths]
    
    unique_patients = list(set(patient_ids))
    n_patients = len(unique_patients)
    
    # Deterministic random shuffle of patients
    np.random.seed(42)
    np.random.shuffle(unique_patients)
    
    n_train = int(round(split_ratios[0] * n_patients))
    n_val = int(round(split_ratios[1] * n_patients))
    
    train_patients_idx = set(unique_patients[:n_train])
    val_patients_idx = set(unique_patients[n_train : n_train+n_val])
    test_patients_idx = set(unique_patients[n_train+n_val :])
    
    train_files, train_labels = [], []
    val_files, val_labels = [], []
    test_files, test_labels = [], []
    
    for idx, pid in enumerate(patient_ids):
        if pid in train_patients_idx:
            train_files.append(filepaths[idx])
            train_labels.append(labels[idx])
        elif pid in val_patients_idx:
            val_files.append(filepaths[idx])
            val_labels.append(labels[idx])
        elif pid in test_patients_idx:
            test_files.append(filepaths[idx])
            test_labels.append(labels[idx])
            
    train_ds = DRDataset(train_files, train_labels, mode)
    val_ds = DRDataset(val_files, val_labels, mode)
    test_ds = DRDataset(test_files, test_labels, mode)
    
    return train_ds, val_ds, test_ds
